lib/stackup/rake_tasks.rb

```diff
@@ -75,42 +78,69 @@ def define
       end
     end
 
-    # Options to "stackup up".
-    #
-    class DataOptions
+    # A flag with optional argument that will be passed to stackup
+    class DataOption
 
-      def initialize
-        @options = {}
+      def initialize(flag, argument)
+        @flag = flag
+        @argument = argument if argument
       end
 
-      def []=(option, file_or_value)
-        @options[option] = file_or_value
-      end
-
-      def files
-        @options.values.grep(String)
+      def to_a
+        [@flag, @argument]
       end
 
-      def to_a
-        [].tap do |result|
-          @options.each do |option, file_or_data|
-            result << option
-            result << maybe_tempfile(file_or_data, option[2..-1])
-          end
+      # Factory method for initialising DataOptions based on class
+      def self.for(flag, argument)
+        case argument
+        when Hash
+          DataOptionHash.new(flag, argument)
+        when Array
+          DataOptionArray.new(flag, argument)
+        when String && File.exist?(argument)
+          DataOptionFile.new(flag, argument)
+        else
+          DataOption.new(flag, argument)
         end
       end
 
-      def maybe_tempfile(file_or_data, type)
-        return file_or_data if file_or_data.is_a?(String)
+    end
+
+    # An option with a Hash argument
+    # Hash content is stored in a temporary file upon conversion
+    class DataOptionHash < DataOption
 
-        tempfile = Tempfile.new([type, ".yml"])
-        tempfile.write(YAML.dump(file_or_data))
+      def as_tempfile(filename, data)
+        tempfile = Tempfile.new(filename)
+        tempfile.write(YAML.dump(data))
         tempfile.close
         tempfile.path.to_s
       end
 
+      def to_a
+        [@flag, as_tempfile([@flag[2..-1], ".yml"], @argument)]
```

Comment:
Since `as_tempfile` is hard coded to deal with YAML files anyway, we may as well move the file extension info into that method.